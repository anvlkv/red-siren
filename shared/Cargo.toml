[package]
name = "shared"
version = "0.1.0"
edition = "2021"
rust-version = { workspace = true }

[lib]
crate-type = ["lib", "staticlib", "cdylib"]
name = "shared"

[features]
typegen = ["crux_core/typegen"]
android = ["android_logger"]
ios = ["oslog"]
worklet = []
browser = ["console_error_panic_hook", "console_log"]

[dependencies]
anyhow = "1.0.71"
bincode = "1.3.3"
crux_core = { workspace = true }
crux_http = { workspace = true }
crux_kv = { workspace = true }
crux_macros = { workspace = true }
crux_platform = { workspace = true }
crux_time = { workspace = true }
derive_more = "0.99.17"
lazy_static = "1.4.0"
serde = { workspace = true, features = ["derive"] }
serde_json = "1.0.96"
thiserror = "1.0.40"
uniffi = "0.25.2"
url = "2.3.1"
wasm-bindgen = "0.2.86"
# project specific
keyframe = "1.1.1"
keyframe_derive = "1.0.0"
mint = { workspace = true, features = ["serde"] }
num-integer = "0.1.45"
hecs = { version = "0.10.3", features = ["serde", "macros"] }
# platforms
cfg-if = "1.0.0"
log = "0.4.17"
android_logger = { version = "0.13.1", optional = true }
oslog = { version = "0.2.0", optional = true }
console_error_panic_hook = { version = "0.1.7", optional = true }
console_log = { version = "1.0.0", optional = true }

[dev-dependencies]
assert_let_bind = "0.1.1"
insta = {workspace = true}

[target.uniffi-bindgen.dependencies]
uniffi = { version = "0.25.2", features = ["cli"] }

[build-dependencies]
uniffi = { version = "0.25.2", features = ["build"] }
