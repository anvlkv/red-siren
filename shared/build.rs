fn main() {
    #[cfg(not(feature = "cargo-clippy"))]
    uniffi::generate_scaffolding("./src/shared.udl").unwrap();
}
