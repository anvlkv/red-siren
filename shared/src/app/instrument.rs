use serde::{Deserialize, Serialize};

use crux_core::render::Render;
use crux_core::App;
use crux_kv::KeyValue;
use crux_macros::Effect;

#[derive(Default)]
pub struct Instrument;

#[derive(Default, Serialize, Deserialize)]
pub struct Model {
    pub config: Config,
}

#[derive(Debug, Default, Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
pub struct Config {
    pub portrait: bool,
    pub width: u16,
    pub height: u16,
    pub groups: u8,
    pub buttons_group: u8,
    pub button_size: u8,
    pub gap: (u8, u8),
    pub steps: u16,
}

impl Config {
    pub fn new(
        width: usize,
        height: usize,
        button_size: usize,
        rigid_step: bool,
        max_buttons: usize,
    ) -> Self {
        let portrait = height > width;
        let (length, breadth) = if portrait {
            (height, width)
        } else {
            (width, height)
        };

        let mut button_gap = button_size / 2;
        let mut button_group_gap = button_gap + button_size;

        let steps = if rigid_step {
            breadth / 2 / button_size
        } else {
            breadth / 2
        };

        let active_length = length - button_group_gap * 2;

        let groups: Vec<(Vec<(usize, usize)>, usize)> = (1..=active_length)
            .into_iter()
            .fold(Vec::<(Vec<(usize, usize)>, usize)>::new(), |mut acc, _| {
                if acc.last().is_some() {
                    let pos = acc.len() - 1;
                    let group = &mut acc[pos];
                    let group_len = group.0.len();

                    let mut last = group.0.last_mut();
                    let btn = last.as_mut().unwrap();
                    if btn.0 < button_size {
                        btn.0 += 1;
                    } else if btn.1 < button_gap && group_len < max_buttons {
                        btn.1 += 1;
                    } else if group_len < max_buttons {
                        group.0.push((1, 0));
                    } else if group.1 < button_group_gap {
                        group.1 += 1;
                    } else {
                        acc.push((vec![(1, 0)], 0));
                    }
                } else {
                    acc.push((vec![(1, 0)], 0));
                }
                acc
            })
            .into_iter()
            .map(|group| {
                (
                    group
                        .0
                        .into_iter()
                        .filter(|btn| btn.0 == button_size)
                        .collect(),
                    group.1,
                )
            })
            .collect();

        let optimal_buttons = groups.iter().map(|g| g.0.len()).max().unwrap_or(1);
        let optimal_groups = groups
            .iter()
            .filter(|g| g.0.len() == optimal_buttons)
            .count();

        if optimal_groups < groups.len() {
            let empty = groups.iter().skip(optimal_groups).fold(0, |acc, group| acc + group.1 + group.0.iter().fold(0, |a, b| a+b.0+b.1));
            let extra = empty / (optimal_groups * optimal_buttons);

            button_gap = button_gap + extra / optimal_groups;
            button_group_gap = button_gap + button_size;
        }

        Config {
            button_size: button_size.try_into().unwrap_or(42),
            portrait,
            width: width.try_into().unwrap_or(200),
            height: height.try_into().unwrap_or(200),
            steps: steps.try_into().unwrap_or(2),
            gap: (
                button_gap.try_into().unwrap_or(24),
                button_group_gap.try_into().unwrap_or(66),
            ),
            groups: optimal_groups.try_into().unwrap_or(1),
            buttons_group: optimal_buttons.try_into().unwrap_or(1),
        }
    }
}

#[derive(Default, Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
pub struct ViewModel {
    pub config: Config,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub enum Event {
    None,
}

#[cfg_attr(feature = "typegen", derive(crux_macros::Export))]
#[derive(Effect)]
#[effect(app = "Instrument")]
pub struct InstrumentCapabilities {
    pub render: Render<Event>,
    pub key_value: KeyValue<Event>,
}

impl App for Instrument {
    type Event = Event;

    type Model = Model;

    type ViewModel = ViewModel;

    type Capabilities = InstrumentCapabilities;

    fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
        todo!()
    }

    fn view(&self, model: &Self::Model) -> Self::ViewModel {
        ViewModel::default()
    }
}
