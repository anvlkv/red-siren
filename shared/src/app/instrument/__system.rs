use super::{
    keyboard::{Button, Track},
    Config, Node,
};
use fundsp::hacker32::*;
use hecs::{Entity, World};

pub struct System {
    pub net: Net32,
    pub net_be: NetBackend32,
    pub size: usize,
    pub entities: Vec<Entity>,
    pub nodes: Vec<NodeId>,
    pub amps: Vec<Shared<f32>>,
    pub tunes: Vec<Shared<f32>>,
}

impl System {
    fn spawn_nodes(world: &mut World) -> Vec<Entity> {
        let mut nodes = world
            .query::<&Button>()
            .iter()
            .map(|(_, b)| {
                let mut query = world.query_one::<&Track>(b.track).unwrap();
                let track = query.get().unwrap();
                (track.freq, b.f_n, if track.left_hand { -1 } else { 1 })
            })
            .collect::<Vec<_>>();

        nodes.sort_by(|a, b| a.1.cmp(&b.1));
        nodes.reverse();

        let nodes = nodes
            .into_iter()
            .map(|(freq, f_n, pan)| Node::spawn(world, freq, f_n, pan))
            .collect::<Vec<_>>();

        nodes
    }

    pub fn spawn(world: &mut World, config: &Config) -> Self {
        let mut net = Net32::new(config.channels, config.channels);
        let entities: Vec<Entity> = System::spawn_nodes(world);
        let size = entities.len();
        let nodes_data = entities
            .iter()
            .map(|e| *world.get::<&Node>(*e).expect("node for entity"))
            .collect::<Vec<_>>();

        let mut amps = Vec::with_capacity(size);
        let mut tunes = Vec::with_capacity(size);
        let mut nodes = Vec::with_capacity(size);

        

        for n in 0..config.channels {
            let c = sine_hz(440.0);
            let id = net.push(Box::new(c));
            nodes.push(id);
            net.connect_output(id, 0, n);
        }
        
        



        net.set_sample_rate(config.sample_rate_hz);

        net.check();

        let mut net_be = net.backend();

        net_be.allocate();

        log::debug!("created network: {}", net.display());

        Self {
            net,
            net_be,
            size,
            entities,
            amps,
            tunes,
            nodes,
        }
    }

    pub fn get_nodes(&self, world: &World) -> Vec<Node> {
        self.entities
            .iter()
            .map(|e| *world.get::<&Node>(*e).expect("node for entity"))
            .collect()
    }
}
