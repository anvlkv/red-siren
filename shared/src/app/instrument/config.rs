use serde::{Deserialize, Serialize};

#[derive(Debug, Default, Serialize, Deserialize, Clone, PartialEq)]
pub struct Config {
    pub portrait: bool,
    pub width: f32,
    pub height: f32,
    pub breadth: f32,
    pub length: f32,
    pub groups: usize,
    pub buttons_group: usize,
    pub button_size: f32,
}

impl Eq for Config {}

impl Config {
    pub fn new(width: f32, height: f32) -> Self {
        let portrait = height > width;

        let (length, breadth) = if portrait {
            (height, width / 3.0)
        } else {
            (width, height / 3.0)
        };

        let bs_size_b = breadth * 0.6;
        let bs_size_l = 932.0 / 6.0 * length / 932.0;

        let button_size = bs_size_b.min(bs_size_l);

        let max_buttons = (length / button_size).floor() as usize;

        let slots = max_buttons.div_euclid(2);

        let groups_2 = slots.div_euclid(2);
        let groups_3 = slots.div_euclid(3);
        let groups_5 = slots.div_euclid(5);

        let (groups, buttons_group) = if groups_5 > 1 {
            (groups_5, 5)
        } else if groups_3 > 1 {
            (groups_3, 3)
        } else {
            (groups_2, 2)
        };

        Config {
            portrait,
            width,
            height,
            length,
            breadth,
            button_size,
            groups: groups.try_into().unwrap_or(1),
            buttons_group,
        }
    }
}
