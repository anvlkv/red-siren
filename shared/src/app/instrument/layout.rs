use super::{
    keyboard::{Button, ButtonGroup, Keyboard, Track},
    string::{InboundString, OutboundString},
};
use crate::geometry::{line::Line, rect::Rect};
use anyhow::Result;
use hecs::{Bundle, Entity, World};
use keyframe::CanTween;
use serde::{Deserialize, Serialize};

#[derive(Default, Serialize, Deserialize, Clone, PartialEq, Debug, Eq)]
pub struct Layout {
    pub inbound: Line,
    pub outbound: Line,
    pub buttons: Vec<Rect>,
    pub tracks: Vec<Rect>,
}

#[derive(Bundle)]
pub struct LayoutRoot {
    entities: (Entity, Entity, Entity),
}

impl LayoutRoot {
    pub fn spawn(world: &mut World, inbound: Entity, outbound: Entity, keyboard: Entity) -> Entity {
        world.spawn((Self {
            entities: (inbound, keyboard, outbound),
        },))
    }

    pub fn inbound(&self) -> Entity {
      self.entities.0
    }
    pub fn outbound(&self) -> Entity {
      self.entities.2
    }
    pub fn keyboard(&self) -> Entity {
      self.entities.1
    }
}

impl CanTween for Layout {
    fn ease(from: Self, to: Self, time: impl keyframe::num_traits::Float) -> Self {
        let outbound = CanTween::ease(from.outbound, to.outbound, time);
        let inbound = CanTween::ease(from.inbound, to.inbound, time);

        let buttons = ease_vec(from.buttons, to.buttons, time);
        let tracks = ease_vec(from.tracks, to.tracks, time);

        Self {
            outbound,
            inbound,
            buttons,
            tracks,
        }
    }
}

impl Layout {
    pub fn new(world: &World, root: &Entity) -> Result<Self> {
        let root = world.get::<&LayoutRoot>(*root)?;
        let (inbound, keyboard, outbound) = (
            world.get::<&InboundString>(root.inbound())?,
            world.get::<&Keyboard>(root.keyboard())?,
            world.get::<&OutboundString>(root.outbound())?,
        );

        let (buttons, tracks) = keyboard
            .groups
            .iter()
            .try_fold(Vec::<(Rect, Rect)>::new(), |mut acc, g| -> Result<_> {
                let group = world.get::<&ButtonGroup>(*g)?;
                for b in &group.buttons {
                    let button = world.get::<&Button>(*b)?;
                    let track = world.get::<&Track>(button.track)?;
                    acc.push((button.rect, track.rect));
                }
                Ok(acc)
            })?
            .into_iter()
            .unzip();

        Ok(Self {
            inbound: inbound.line,
            outbound: outbound.line,
            buttons,
            tracks,
        })
    }

    pub fn dummy(breadth: f64, length: f64, button_size: f64) -> Self {
        let inbound = Line::new(0.0, length, 0.0, 1.0);
        let outbound = Line::new(0.0, length, breadth, breadth + 1.0);

        let button = Rect::new(0.0, button_size, 0.0, button_size);

        Self {
            inbound,
            outbound,
            buttons: vec![button],
            ..Default::default()
        }
    }
}

fn ease_vec(from: Vec<Rect>, to: Vec<Rect>, time: impl keyframe::num_traits::Float) -> Vec<Rect> {
    let len = CanTween::ease(from.len() as f64, to.len() as f64, time).round() as usize;
    (0..len)
        .into_iter()
        .map(|i| {
            let (from_b, to_b) = (from.iter().nth(i), to.iter().nth(i));
            match (from_b, to_b) {
                (Some(from), Some(to)) => CanTween::ease(*from, *to, time),
                (Some(from), None) => {
                    let c = from.center();
                    let z_rect = Rect::new(c.x, c.x, c.y, c.y);
                    CanTween::ease(*from, z_rect, time)
                }
                (None, Some(to)) => {
                    let c = to.center();
                    let z_rect = Rect::new(c.x, c.x, c.y, c.y);
                    CanTween::ease(z_rect, *to, time)
                }
                (None, None) => unreachable!(),
            }
        })
        .collect()
}
