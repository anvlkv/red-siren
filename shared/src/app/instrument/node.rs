use serde::{Deserialize, Serialize};

#[derive(Default, Serialize, Deserialize, Clone, PartialEq)]
pub struct Node {
    pub base_freq: f64,
    pub max_freq: f64,
    pub f_n: usize,
}

impl Eq for Node {}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        match self.base_freq.partial_cmp(&other.base_freq) {
            Some(core::cmp::Ordering::Equal) => {}
            ord => return ord,
        }
        self.max_freq.partial_cmp(&other.max_freq)
    }
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.partial_cmp(other).unwrap_or(std::cmp::Ordering::Equal)
    }
}

impl Node {
    pub fn new((base_freq, max_freq): (f64, f64), f_n: usize) -> Self {
        Self {
            base_freq,
            max_freq,
            f_n,
        }
    }
}
