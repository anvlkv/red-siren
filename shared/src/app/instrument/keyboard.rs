use super::config::Config;
use crate::geometry::Rect;
use hecs::{Bundle, Entity, World};

#[derive(Bundle)]
pub struct Track {
    pub left_hand: bool,
    pub rect: Rect,
}

impl Track {
    pub fn spawn(
        world: &mut World,
        config: &Config,
        left_hand: bool,
        button_rect: &Rect,
    ) -> Entity {
        let sign = if left_hand { -1.0 } else { 1.0 };
        let button_track_margin = config.button_size * 0.2 * sign;
        let track_length = (config.breadth * 2.0 + button_track_margin * sign + config.button_size) * sign;
        let track_breadth = (config.button_size + button_track_margin * 2.0 * sign) * sign;

        let side = if left_hand {
            button_rect.bottom_right()
        } else {
            button_rect.top_left()
        };

        let rect = if config.portrait {
            Rect::new(
                side.x - button_track_margin,
                side.x - button_track_margin + track_length,
                side.y - button_track_margin,
                side.y - button_track_margin + track_breadth,
            )
        } else {
            Rect::new(
                side.x - button_track_margin,
                side.x - button_track_margin + track_breadth,
                side.y - button_track_margin,
                side.y - button_track_margin + track_length,
            )
        };

        world.spawn((Self { rect, left_hand },))
    }
}

#[derive(Bundle)]
pub struct Button {
    pub track: Entity,
    pub rect: Rect,
    pub group_button: (usize, usize),
}

impl Button {
    pub fn spawn(world: &mut World, config: &Config, group: usize, button: usize) -> Entity {
        let button_space_b = (config.breadth - config.button_size) / 2.0;
        let button_space_l = (config.length / (config.groups * config.buttons_group) as f32
            - config.button_size)
            / 2.0;
        let side = config.breadth + button_space_b;
        let side_breadth = side + config.button_size;
        let idx = (group - 1) * config.buttons_group + (button - 1);
        let main = (config.button_size + button_space_l * 2.0) * idx as f32 + button_space_l;
        let main_length = main + config.button_size;
        let rect = if config.portrait {
            Rect::new(side, side_breadth, main, main_length)
        } else {
            Rect::new(main, main_length, side, side_breadth)
        };

        let track = Track::spawn(world, config, group % 2 == 0, &rect);
        world.spawn((Button {
            rect,
            track,
            group_button: (group, button),
        },))
    }
}

#[derive(Bundle)]
pub struct ButtonGroup {
    pub buttons: Vec<Entity>,
    pub rect: Rect,
}

impl ButtonGroup {
    pub fn spawn(world: &mut World, config: &Config, group: usize) -> Entity {
        let mut buttons = vec![];
        for j in 1..=config.buttons_group {
            buttons.push(Button::spawn(world, config, group, j));
        }
        let group_length = config.length / config.groups as f32;
        let rect = if config.portrait {
            Rect::new(
                config.breadth,
                config.breadth * 2.0,
                group_length * (group - 1) as f32,
                group_length * group as f32,
            )
        } else {
            Rect::new(
                group_length * (group - 1) as f32,
                group_length * group as f32,
                config.breadth,
                config.breadth * 2.0,
            )
        };

        world.spawn((Self { buttons, rect },))
    }
}

#[derive(Bundle)]
pub struct Keyboard {
    pub groups: Vec<Entity>,
    pub rect: Rect,
}

impl Keyboard {
    pub fn spawn(world: &mut World, config: &Config) -> Entity {
        let mut groups = vec![];
        for i in 1..=config.groups {
            groups.push(ButtonGroup::spawn(world, config, i));
        }
        let rect = Rect::size(config.width, config.height);
        world.spawn((Keyboard { groups, rect },))
    }
}
