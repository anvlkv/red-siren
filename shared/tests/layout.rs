use hecs::World;
use shared::instrument::{keyboard, string, Config, Layout, LayoutRoot};

#[test]
fn creates_layout_from_config() {
    let config = Config::new(430.0, 932.0, 476.0, Default::default());
    let mut world = World::new();

    let inbound = string::InboundString::spawn(&mut world, &config);
    let outbound = string::OutboundString::spawn(&mut world, &config);
    let keyboard = keyboard::Keyboard::spawn(&mut world, &config);

    let root = LayoutRoot::spawn(&mut world, inbound, outbound, keyboard);
    let layout = Layout::new(&world, &root).expect("failed to create layout");

    insta::assert_yaml_snapshot!(layout)
}
