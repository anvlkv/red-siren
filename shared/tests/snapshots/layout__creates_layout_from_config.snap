---
source: shared/tests/layout.rs
expression: layout
---
inbound:
  line:
    - - 143.33334
      - 0
    - - 143.33334
      - 932
outbound:
  line:
    - - 286.6667
      - 0
    - - 286.6667
      - 932
buttons:
  - rect:
      - - 172
        - 73.5
      - - 258
        - 159.5
  - rect:
      - - 172
        - 306.5
      - - 258
        - 392.5
  - rect:
      - - 172
        - 539.5
      - - 258
        - 625.5
  - rect:
      - - 172
        - 772.5
      - - 258
        - 858.5
tracks:
  - rect:
      - - 154.8
        - 56.3
      - - 544.6667
        - 176.7
  - rect:
      - - 154.8
        - 289.3
      - - 544.6667
        - 409.69998
  - rect:
      - - -114.66666
        - 522.3
      - - 275.2
        - 642.7
  - rect:
      - - -114.66666
        - 755.3
      - - 275.2
        - 875.7

